### ISSUE TITLE
Error message in notebook when running Explicit Path graph example from User Guide

### ISSUE THREAD

**Issue by hsparra** [bokeh/bokeh#7011]
When I take the Explicit Plot example in the graph section of the User Guide, https://bokeh.pydata.org/en/latest/docs/user_guide/graph.html, and display in a notebook I get the following message on the `show(plot)`:
`
E-1001 (BAD_COLUMN_NAME): Glyph refers to nonexistent column name: fill_color [renderer: GlyphRenderer(id='2ee90517-57dc-45e6-8169-288256a4d6cf', …)]
`

The plot still displays as expected.

#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
Bokeh: 0.12.9
Python: 3.6

#### Description of expected behavior and the observed behavior
Plot displays as it does now, but the error message does not display.

#### Complete, minimal, self-contained example code that reproduces the issue

```
import math

from bokeh.io import show, output_notebook
from bokeh.plotting import figure
from bokeh.models import GraphRenderer, StaticLayoutProvider, Oval
from bokeh.palettes import Spectral8

output_notebook()

N = 8
node_indices = list(range(N))

plot = figure(title="Graph Layout Demonstration", x_range=(-1.1,1.1), y_range=(-1.1,1.1),
              tools="", toolbar_location=None)

graph = GraphRenderer()

graph.node_renderer.data_source.data = dict(
    index=node_indices,
    fill_color=Spectral8)
graph.node_renderer.glyph = Oval(height=0.1, width=0.2, fill_color="fill_color")

graph.edge_renderer.data_source.data = dict(
    start=[0]*N,
    end=node_indices)

### start of layout code
circ = [i*2*math.pi/8 for i in node_indices]
x = [math.cos(i) for i in circ]
y = [math.sin(i) for i in circ]
graph_layout = dict(zip(node_indices, zip(x, y)))
graph.layout_provider = StaticLayoutProvider(graph_layout=graph_layout)

### Draw quadratic bezier paths
def bezier(start, end, control, steps):
    return [(1-s)**2*start + 2*(1-s)*s*control + s**2*end for s in steps]

xs, ys = [], []
sx, sy = graph_layout[0]
steps = [i/100. for i in range(100)]
for node_index in node_indices:
    ex, ey = graph_layout[node_index]
    xs.append(bezier(sx, ex, 0, steps))
    ys.append(bezier(sy, ey, 0, steps))
graph.edge_renderer.data_source.data['xs'] = xs
graph.edge_renderer.data_source.data['ys'] = ys

plot.renderers.append(graph)

show(plot)
```

#### Stack traceback and/or browser JavaScript console output
None

#### Screenshots or screencasts of the bug in action
![06_explicit_path_edges_bokeh 2017-10-04 12-11-22](https://user-images.githubusercontent.com/7515105/31189253-46c863a8-a8fd-11e7-842f-f205c4ab2cb3.png)

**Comment by mattpap** [comment]
The problem is with field name validator in:
https://github.com/bokeh/bokeh/blob/bbdf7bd63a255da7a99b7583bc4846684d6163fb/bokeh/models/renderers.py#L116-L128

`self.data_source.column_names` is not automatically updated when `data` changes.

**Comment by mattpap** [comment]
@bryevdv, I'm not sure, but I think this should use `self.data_source.data.keys()` instead?

**Comment by bryevdv** [comment]
possibly I'll take a look soon

**Comment by David-Herman** [comment]
Hello,

I receive the same error when running bokeh server and running the above code but with the addition of 
```
doc = curdoc()  
doc.add_root(plot)
```
And then running the serve --show command.

My use case is that I would like to dynamically update the node and edge source data to denote information, such as shortest path between two nodes via a python callback using networkx. So the inability to update the graph is an issue for my application.

**Comment by viveksarmar** [comment]
@mattpap - Restarting ipython Kernel in my Anaconda flushed the old data Columns.

**Comment by hsparra** [comment]
@David-Herman, at least as of bokeh 0.12.13 you can update the graph using python callbacks. Your update would be something like:

```

**Comment by mattpap** [comment]
@bryevdv, wasn't this resolved by PR #7740?

**Comment by bryevdv** [comment]
Seems to be!

<img width="601" alt="screen shot 2018-04-08 at 14 45 14" src="https://user-images.githubusercontent.com/1078448/38471772-7b39aff4-3b3b-11e8-9621-ce9df67adc4b.png">
